[Wasm-GC] Implement cast operations

takikawa opened the pull request:
#### 0e2cbf8c3d8ee53e2840c3b87aa7b0adaf410865
<pre>
[Wasm-GC] Implement cast operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=247399">https://bugs.webkit.org/show_bug.cgi?id=247399</a>

Reviewed by Yusuke Suzuki.

Implements the `ref.cast` and `ref.test` instructions. Casts are
supported by associating an RTT with Wasm objects.

For functions, this is done by adding a field in the base function
object.  For arrays and structs, a new WebAsseblyGCObjectBase class is
used as the base class for both arrays and structs. The base class
contains an RTT with a consistent offset for the JIT.

This patch does not yet support `br_on_cast/on_cast_fail`.

* JSTests/wasm/gc/casts.js: Added.
(testBasicCasts):
(testI31Casts):
(testFunctionCasts):
(testArrayCasts):
(testStructCasts):
(testSubtypeCasts):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator32_64.cpp:
(JSC::Wasm::AirIRGenerator32::emitBranchForNullReference):
(JSC::Wasm::AirIRGenerator32::makeBranchNotInt32):
(JSC::Wasm::AirIRGenerator32::makeBranchNotCell):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::emitBranchForNullReference):
(JSC::Wasm::AirIRGenerator64::makeBranchNotInt32):
(JSC::Wasm::AirIRGenerator64::makeBranchNotCell):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::addRefCast):
(JSC::Wasm::ExpressionType&gt;::addRefTest):
(JSC::Wasm::ExpressionType&gt;::emitRefTestOrCast):
(JSC::Wasm::ExpressionType&gt;::emitCheckOrBranchForCast):
(JSC::Wasm::ExpressionType&gt;::emitLoadRTTFromFuncref):
(JSC::Wasm::ExpressionType&gt;::emitLoadRTTFromObject):
(JSC::Wasm::ExpressionType&gt;::makeBranchNotRTTKind):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addRefTest):
(JSC::Wasm::B3IRGenerator::addRefCast):
(JSC::Wasm::B3IRGenerator::emitRefTestOrCast):
(JSC::Wasm::B3IRGenerator::emitCheckOrBranchForCast):
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromFuncref):
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromObject):
(JSC::Wasm::B3IRGenerator::emitNotRTTKind):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::initElementSegment):
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addRefTest):
(JSC::Wasm::LLIntGenerator::addRefCast):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::arrayNew):
(JSC::Wasm::structNew):
(JSC::Wasm::refCast):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::RTT::tryCreateRTT):
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
(JSC::Wasm::TypeInformation::canonicalRTTForType):
(JSC::Wasm::TypeInformation::tryGetCanonicalRTT):
(JSC::Wasm::TypeInformation::getCanonicalRTT):
(): Deleted.
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::RTT::RTT): Deleted.
(JSC::Wasm::RTT::displaySize const): Deleted.
(JSC::Wasm::RTT::displayEntry const): Deleted.
(JSC::Wasm::RTT::setDisplayEntry): Deleted.
(JSC::Wasm::RTT::allocatedRTTSize): Deleted.
(JSC::Wasm::RTT::payload): Deleted.
* Source/JavaScriptCore/wasm/WasmTypeDefinitionInlines.h:
(JSC::Wasm::TypeInformation::getFunctionSignature):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
(JSC::JSWebAssemblyArray::JSWebAssemblyArray):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
(JSC::JSWebAssemblyStruct::tryCreate):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::create):
(JSC::WebAssemblyFunction::WebAssemblyFunction):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.cpp:
(JSC::WebAssemblyFunctionBase::WebAssemblyFunctionBase):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::rtt const):
(JSC::WebAssemblyFunctionBase::offsetOfRTT):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.cpp: Copied from Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h.
(JSC::WebAssemblyGCObjectBase::WebAssemblyGCObjectBase):
(JSC::WebAssemblyGCObjectBase::visitChildrenImpl):
(JSC::WebAssemblyGCObjectBase::finishCreation):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h: Copied from Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h.
(JSC::WebAssemblyGCObjectBase::rtt):
(JSC::WebAssemblyGCObjectBase::offsetOfRTT):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.cpp:
(JSC::WebAssemblyWrapperFunction::WebAssemblyWrapperFunction):
(JSC::WebAssemblyWrapperFunction::create):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/261445@main">https://commits.webkit.org/261445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0c6c69f0f7ffa883be1f20c524f27f106d7d3f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111597 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/202 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3471 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120349 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11812 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117362 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99555 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104491 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/133 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45271 "") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100101 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13217 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/130 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11339 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9570 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101316 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19156 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52118 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31524 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7964 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15694 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109355 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26927 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

takikawa commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -876,6 +894,8 @@ class TypeInformation {
     RefPtr<TypeDefinition> m_Void_I32I32I32I32;
     RefPtr<TypeDefinition> m_Void_I32I32I32I32I32;
     RefPtr<TypeDefinition> m_I32_I32;
+    RefPtr<TypeDefinition> m_I32_RefI32I32;
```
Note: the style checker complains about these class members, but they are named to be consistent with the other types above.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2972,6 +2982,193 @@ auto B3IRGenerator::addStructSet(ExpressionType structReference, const StructTyp
+Value* B3IRGenerator::emitLoadRTTFromObject(Value* reference)
+{
+    PatchpointValue* patch = m_currentBlock->appendNew<PatchpointValue>(m_proc, B3::Int64, Origin());
+    patch->clobber(RegisterSetBuilder::macroClobberedGPRs());
```
I'm assuming these effects are right, but we might want a second pair of eyes

justinmichaud commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -113,11 +114,13 @@ auto SectionParser::parseType() -> PartialResult
                     RefPtr<TypeDefinition> group = TypeInformation::typeDefinitionForRecursionGroup(types);
                     RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(group->index(), 0);
                     TypeInformation::registerCanonicalRTTForType(projection->index());
+                    m_info->rtts.uncheckedAppend(TypeInformation::getCanonicalRTT(projection->index()));
```
Does this need a lock?

Constellation replied:
Unnecessary since section-parser is single-thread. (FunctionParser is concurrent)

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
```diff
@@ -97,11 +97,6 @@ void JSWebAssemblyArray::finishCreation(VM& vm)
 {
     Base::finishCreation(vm);
     ASSERT(inherits(info()));
-
```
but we didn't change the JSWebAssemblyArray supertype, right? So isn't this wrong?

takikawa replied:
This is about the `seal` right? The issue is that this doesn't work without structure transitions, which the wasm structure in this patch doesn't support. The `seal` is also unnecessary with the JS API patch (https://github.com/WebKit/WebKit/pull/10408), so it was removed here assuming that other patch would also land.

That said, if we apply @Constellation's suggestion to put the RTT in the object instance, it's not necessary to remove this line in this patch.

justinmichaud commented on Source/JavaScriptCore/wasm/js/WebAssemblyGCStructure.cpp:
```diff
@@ -0,0 +1,60 @@
+
+namespace JSC {
+
+WebAssemblyGCStructure::WebAssemblyGCStructure(VM& vm, JSGlobalObject* globalObject, JSValue prototype, const TypeInfo& typeInfo, const ClassInfo* classInfo, RefPtr<const Wasm::RTT> rtt)
```
Someone else needs to review this file, I don't know enough

justinmichaud reviewed (commented):
Overall LGTM. I am assuming the semantics are implemented correctly based on the tests. I would r+, but I think someone else should take a quick second look some of the more-involved b3 and JSObject changes.

Constellation commented on Source/JavaScriptCore/runtime/Structure.cpp:
```diff
@@ -340,6 +341,11 @@ Structure::~Structure()
+#if ENABLE(WEBASSEMBLY)
+    if (isWebAssemblyGCStructure())
+        static_cast<WebAssemblyGCStructure*>(this)->destruct();
+#endif
```
Why don't we put RTT object directly into instance instead of Structure? Extending Structure has ***super*** high bar. If the other method works, then we should avoid it.

takikawa replied:
This is now done, and I've left some comments explaining how the code changed to make this work.

Constellation reviewed (changes requested):
Left comment about RTT in Structure. I think putting RTT on instance is better than extending Structure.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyGCStructure.h:
```diff
@@ -0,0 +1,68 @@
+
+    CompactRefPtr<const Wasm::RTT> rtt() const { return m_rtt; }
+
+    static ptrdiff_t offsetOfRTT() { return OBJECT_OFFSETOF(WebAssemblyGCStructure, m_rtt); }
```
Also, we should not assume that this will not get transition. Structure transition occurs various reasons, like, prototype change, etc. And there is no contract which prevents it completely in JSC. So, assuming structure transition does not happen is not working. This is another reason why we should put RTT in instance instead of extending structure.

takikawa commented on Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h:
```diff
@@ -0,0 +1,57 @@
+
+namespace JSC {
+
+class WebAssemblyGCObjectBase : public JSNonFinalObject {
```
I added this new base class in the latest revision. It's used as the base class for arrays and structs in order to store an RTT field with a consistent offset for the JIT.

It'll actually be useful in the follow-up JS API patch, as the object methods can probably be defined on this.

takikawa commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -192,6 +195,59 @@ inline void structSet(Instance* instance, EncodedJSValue encodedStructReference,
+    default: {
+        ASSERT(!Wasm::typeIndexIsType(static_cast<Wasm::TypeIndex>(heapType)));
+        Wasm::TypeDefinition& signature = instance->module().moduleInformation().typeSignatures[heapType];
+        auto signatureRTT = instance->module().moduleInformation().rtts[heapType];
```
We just always use the RTT from the module information now, as we don't store a list of Structures in the Wasm Module now.

takikawa commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -192,6 +195,59 @@ inline void structSet(Instance* instance, EncodedJSValue encodedStructReference,
+            JSWebAssemblyArray* arrayRef = jsDynamicCast<JSWebAssemblyArray*>(refValue);
+            if (!arrayRef)
+                return false;
+            auto arrayRTT = arrayRef->rtt();
```
For structs and arrays this just accesses the RTT from the object directly (before it looked up the Structure, then RTT in the structure)

takikawa commented on Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
```diff
@@ -2695,6 +2706,185 @@ auto AirIRGeneratorBase<Derived, ExpressionType>::addStructSet(ExpressionType st
+template <typename Derived, typename ExpressionType>
+void AirIRGeneratorBase<Derived, ExpressionType>::emitLoadRTTFromObject(ExpressionType cell, ExpressionType& result)
+{
+    self().emitLoad(cell, WebAssemblyGCObjectBase::offsetOfRTT(), result);
```
Both Air and B3 implementations for this got simpler, as it's now just a plain load instead of a patchpoint.

takikawa commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -54,9 +54,11 @@ inline EncodedJSValue arrayNew(Instance* instance, uint32_t typeIndex, uint32_t
     const Wasm::TypeDefinition& arraySignature = instance->module().moduleInformation().typeSignatures[typeIndex]->expand();
     ASSERT(arraySignature.is<ArrayType>());
     Wasm::FieldType fieldType = arraySignature.as<ArrayType>()->elementType();
+    auto arrayRTT = instance->module().moduleInformation().rtts[typeIndex];
```
The RTT now has to be supplied on object construction right here, rather than when the structure is created. This isn't a problem because all the RTTs are tracked in the module information.

Constellation reviewed (approved):
r=me Please reupload a patch to ensure that all tests pass (since recently BBQJIT turned on by default).

takikawa committed:
[Wasm-GC] Implement cast operations
https://bugs.webkit.org/show_bug.cgi?id=247399

Reviewed by Yusuke Suzuki.

Implements the `ref.cast` and `ref.test` instructions. Casts are
supported by associating an RTT with Wasm objects.

For functions, this is done by adding a field in the base function
object.  For arrays and structs, a new WebAsseblyGCObjectBase class is
used as the base class for both arrays and structs. The base class
contains an RTT with a consistent offset for the JIT.

This patch does not yet support `br_on_cast/on_cast_fail`.

* JSTests/wasm/gc/casts.js: Added.
(testBasicCasts):
(testI31Casts):
(testFunctionCasts):
(testArrayCasts):
(testStructCasts):
(testSubtypeCasts):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator32_64.cpp:
(JSC::Wasm::AirIRGenerator32::emitBranchForNullReference):
(JSC::Wasm::AirIRGenerator32::makeBranchNotInt32):
(JSC::Wasm::AirIRGenerator32::makeBranchNotCell):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::emitBranchForNullReference):
(JSC::Wasm::AirIRGenerator64::makeBranchNotInt32):
(JSC::Wasm::AirIRGenerator64::makeBranchNotCell):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::addRefCast):
(JSC::Wasm::ExpressionType>::addRefTest):
(JSC::Wasm::ExpressionType>::emitRefTestOrCast):
(JSC::Wasm::ExpressionType>::emitCheckOrBranchForCast):
(JSC::Wasm::ExpressionType>::emitLoadRTTFromFuncref):
(JSC::Wasm::ExpressionType>::emitLoadRTTFromObject):
(JSC::Wasm::ExpressionType>::makeBranchNotRTTKind):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addRefTest):
(JSC::Wasm::B3IRGenerator::addRefCast):
(JSC::Wasm::B3IRGenerator::emitRefTestOrCast):
(JSC::Wasm::B3IRGenerator::emitCheckOrBranchForCast):
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromFuncref):
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromObject):
(JSC::Wasm::B3IRGenerator::emitNotRTTKind):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::initElementSegment):
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addRefTest):
(JSC::Wasm::LLIntGenerator::addRefCast):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::arrayNew):
(JSC::Wasm::structNew):
(JSC::Wasm::refCast):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::RTT::tryCreateRTT):
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
(JSC::Wasm::TypeInformation::canonicalRTTForType):
(JSC::Wasm::TypeInformation::tryGetCanonicalRTT):
(JSC::Wasm::TypeInformation::getCanonicalRTT):
(): Deleted.
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::RTT::RTT): Deleted.
(JSC::Wasm::RTT::displaySize const): Deleted.
(JSC::Wasm::RTT::displayEntry const): Deleted.
(JSC::Wasm::RTT::setDisplayEntry): Deleted.
(JSC::Wasm::RTT::allocatedRTTSize): Deleted.
(JSC::Wasm::RTT::payload): Deleted.
* Source/JavaScriptCore/wasm/WasmTypeDefinitionInlines.h:
(JSC::Wasm::TypeInformation::getFunctionSignature):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp:
(JSC::JSWebAssemblyArray::JSWebAssemblyArray):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
(JSC::JSWebAssemblyStruct::tryCreate):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::create):
(JSC::WebAssemblyFunction::WebAssemblyFunction):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.cpp:
(JSC::WebAssemblyFunctionBase::WebAssemblyFunctionBase):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::rtt const):
(JSC::WebAssemblyFunctionBase::offsetOfRTT):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.cpp: Copied from Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h.
(JSC::WebAssemblyGCObjectBase::WebAssemblyGCObjectBase):
(JSC::WebAssemblyGCObjectBase::visitChildrenImpl):
(JSC::WebAssemblyGCObjectBase::finishCreation):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h: Copied from Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h.
(JSC::WebAssemblyGCObjectBase::rtt):
(JSC::WebAssemblyGCObjectBase::offsetOfRTT):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.cpp:
(JSC::WebAssemblyWrapperFunction::WebAssemblyWrapperFunction):
(JSC::WebAssemblyWrapperFunction::create):
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/261445@main
